pennylane/ops/qubit/matrix_ops.py

```diff
@@ -16,13 +16,139 @@
 accept a hermitian or an unitary matrix as a parameter.
 """
 # pylint:disable=arguments-differ
+from functools import lru_cache, reduce
+from itertools import product
 import warnings
 import numpy as np
 
 import pennylane as qml
 from pennylane.operation import AnyWires, DecompositionUndefinedError, Operation
 from pennylane.wires import Wires
 
+_paulis = np.array([[[1, 0], [0, 1]], [[0, 1], [1, 0]], [[0, -1j], [1j, 0]], [[1, 0], [0, -1]]])
+"""Single-qubit Paulis:    I                 X                   Y                  Z"""
+
+_pauli_letters = ["I", "X", "Y", "Z"]
+"""Single-qubit Pauli letters that make up Pauli words."""
+
+
+@lru_cache
```

Comment:
Do we want to be performing caching on something potentially extremely memory heavy? Can we use a cache with a memory limit so we don't store large matrices after we don't use them anymore?

Reply:
Good call! My plan for the future would be that we do not create this massive tensor if the number of wires becomes very large, but we iteratively construct the special unitary matrix in `special_unitary_matrix` instead.

In this case, PL itself would not call `pauli_basis` for very large numbers of wires anyways and we could point users to not doing so either.

Do you think this is something that should be taken care of in this PR?

Reply:
I played around with a prototype to mix the two approaches, but come to the conclusion that completely separating them is nicer to avoid too large tensors being created intermediately.
I committed a version of `special_unitary_matrix` that uses  `pauli_basis` only up to five wires. We could make that number six, I'm not sure how much memory is considered much.

Reply:
I now committed a version that uses `pauli_basis` for up to five wires, and a sequential generation process afterwards. There is definitely potential for optimization for `num_wires>6`, but I don't want to go into premature optimization here.

Reply:
Given the limited scope of the operation and the failsafe switch to not use the dense tensor `pauli_basis_matrices` in `SpecialUnitary.compute_matrix` for very large dimensions, I think this is fine now. I'll resolve.